data_steward/analytics/cross_ct_rt/CDR_QC_RT_vs_CT.py

```diff
@@ -0,0 +1,497 @@
+from omop o
+join ext e 
+on o.column_name = e.column_name and o.table_name != e.table_name
+where REGEXP_CONTAINS(e.table_name, o.table_name)
+order by table_name
+)
+
+SELECT AS STRUCT table_name, column_name
+FROM data
+);
+
+CREATE TEMP TABLE result(table_name STRING, in_rt_but_not_ct INT64, Failure INT64);
+
+LOOP
+  SET i = i + 1;
+  IF i > ARRAY_LENGTH(omop_tables) THEN 
+    LEAVE; 
+  END IF;
+  EXECUTE IMMEDIATE '''
+INSERT result
+select 
+"''' || omop_tables[ORDINAL(i)].table_name || '''" as table_name,
+count(*) as in_rt_but_not_ct,
+CASE WHEN 
+  count(*) > 0
+  THEN 1 ELSE 0
+END
+ as Failure
+FROM `{project_id}.{rt_dataset}.''' || omop_tables[ORDINAL(i)].table_name || '''` r
+where ''' || omop_tables[ORDINAL(i)].column_name || ''' not in (
+SELECT ''' || omop_tables[ORDINAL(i)].column_name || ''' 
+FROM `{project_id}.{ct_dataset}.''' || omop_tables[ORDINAL(i)].table_name || '''`)
+''';
+
+END LOOP;
+
+-- success is calculated per table in the temporary table --
+SELECT *
+FROM result
+ORDER BY Failure DESC, table_name;
+"""
+
+df1=pd.read_gbq(query, dialect='standard')  
+
+
+if df1['Failure'].sum()==0:
+ df = df.append({'query' : 'Query5 Verify All primary keys in RT are available in CT', 'result' : 'PASS'},  
+                ignore_index = True) 
+else:
+ df = df.append({'query' : 'Query5 Verify All primary keys in RT are available in CT' , 'result' : ''},  
+                ignore_index = True) 
+
+df1[df1["Failure"] ==1]
+
+# -
+
+# # Verify All person_ids in RT exist in CT
+# Run a check to determine if any person_ids exist in the RT that
+# do not exist in the CT.  If this happens, we suspect an error exists.
+# This is a signal that something is wrong, it will be hard to determine
+# if the issue is with RT or CT code without reviewing the github
+# changes between the last known functional CDR and the current
+# erroneous CDR.  If count is 0, no issues were encountered.
+
+
+# +
+query = f"""
+DECLARE person_tables Array<STRING>;
+DECLARE i INT64 DEFAULT 0;
+
+set person_tables = Array(
+    SELECT table_name                                                                                                                                                                                
+    FROM `{project_id}.{rt_dataset}.INFORMATION_SCHEMA.COLUMNS`
+    where lower(column_name) = 'person_id'
+);
+
+CREATE TEMP TABLE rt_person_id(person_id INT64);
+CREATE TEMP TABLE ct_person_id(person_id INT64);
+
+LOOP
+  SET i = i + 1;
+  IF i > ARRAY_LENGTH(person_tables) THEN
+    LEAVE;
+  END IF;
+  EXECUTE IMMEDIATE '''
+INSERT rt_person_id
+SELECT distinct person_id
+FROM `{project_id}.{rt_dataset}.''' || person_tables[ORDINAL(i)] || '''`
+''';
+
+END LOOP;
+
+set i = 0;
+LOOP
+  SET i = i + 1;
+  IF i > ARRAY_LENGTH(person_tables) THEN
+    LEAVE;
+  END IF;
+  EXECUTE IMMEDIATE '''
+INSERT ct_person_id
+SELECT distinct person_id
+FROM `{project_id}.{ct_dataset}.''' || person_tables[ORDINAL(i)] || '''`
+''';
```

Comment:
This looks like it is appending person_ids from every table into a single table. There can be a lot of repeated person_ids from different tables since the distinct is limited to the current table. I believe these can be simplified to not use a loop, but we can discuss more if you have time tomorrow. Let me know your thoughts.

Reply:
@lrwb-aou @nishanthpp93  Lauren will look into this. thanks.

Reply:
@nishanthpp93 What do you suggest?  This was done in a hurry and could possibly be optimized.  I will document my initial decisions for these queries here so we can discuss.
1. I didn't want to hard code all the tables with `person_id` as a column and wanted to read this from the available tables in the RT dataset since that is the target dataset for this verification.
2. I wanted to get all person_ids without reading them from the `person` table because an error affecting the `person` table may hide an error with the RT/CT cross dataset verification.
3. I did try to limit to reading the distinct `person_id`s for each OMOP table to make this a less expensive process.  I know there will be some duplicates due the person_id existing in multiple tables.  I felt this was ok because it was, at most, #_of_participants * #_of_tables_with_personid_column.  These tables are also not being stored, but are temporary.
4. I think line 403 should be changed to `WHERE person_id not in (select distinct person_id from ct_person_id);`

Reply:
@lrwb-aou thanks for the context. I think 4 should definitely help. Do you think line 396 can have the same where clause while appending to the table?

Reply:
@nishanthpp93 I think that could work.  It could be done on 396 and 381.  But we would want to check to make sure it is not already in the temporary table instead of the other tier table.

Reply:
@nishanthpp93 @lrwb-aou I am not sure what exactly we have to do about line 396 and 381, it will be helpful if this can be finalized. thanks.

Reply:
@jqian2015 This would be adding
```
WHERE person_id not in (select distinct person_id from rt_person_id)
``` 
to the `INSERT rt_person_id` SQL and
```
WHERE person_id not in (select distinct person_id from ct_person_id)
``` 
to the `INSERT ct_person_id` SQL.  The statement should be added as an additional line before the line containing `''';`

Reply:
will do. thanks.